### PR TITLE
Implement the Module cloning functionality

### PR DESCRIPTION
**Pull request by opti-mix**
Follow the examples of the existing cloning APIs for nodes and functions and extend this functionality to work with modules as well.

While doing that, also tighten the graph verifier checks when it comes to types used inside a module.

### CODE REVIEW

File: tests/unittests/GraphTest.cpp [hunk 1]
```diff
@@ -674,6 +674,89 @@ TEST(Graph, functionCloneTest) {
...
+    auto *F = originalM.createFunction("main");
+    Node *input1 = originalM.createPlaceholder(ElemKind::FloatTy,
+                                               {4, 10, 10, 3}, "input", true);
+
+    Node *add = F->createAdd("add", input1, input1);
+    Node *relu = F->createRELU("Relu", add);
+    auto *concat = F->createConcat("concat", {relu, relu, relu}, 0);
+    auto *C = originalM.createConstant(concat->getResult().getType(), "C");
+    C->getPayloadMutable().getHandle().clear(1.0f);
+    Node *SM = F->createAdd("sub", concat, C);
```
Comment (jackm321) [inline tests/unittests/GraphTest.cpp]: nit "add"

File: lib/Graph/Graph.cpp [hunk 2]
```diff
@@ -4093,6 +4123,43 @@ static bool verifyNodeInput(const Node &N, size_t idx) {
...
+  for (auto &C : getConstants()) {
+    auto *copyC = M->createConstant(C->getType(), C->getName(), C->getLayout());
+    copyC->assign(&C->getPayload());
+    currToNew[C] = copyC;
+  }
+  // Clone all functions.
+  for (auto &F : getFunctions()) {
+    // Create an empty clone function in the new module.
+    auto copyF = M->createFunction(F->getName());
+    // Clone function's body into the newy created cloned function. Use the
```
Comment (shajrawi) [inline lib/Graph/Graph.cpp]: nit: newly

File: lib/Graph/Graph.cpp [hunk 3]
```diff
@@ -4093,6 +4123,43 @@ static bool verifyNodeInput(const Node &N, size_t idx) {
...
+  // Clone constants.
+  for (auto &C : getConstants()) {
+    auto *copyC = M->createConstant(C->getType(), C->getName(), C->getLayout());
+    copyC->assign(&C->getPayload());
+    currToNew[C] = copyC;
+  }
+  // Clone all functions.
+  for (auto &F : getFunctions()) {
+    // Create an empty clone function in the new module.
+    auto copyF = M->createFunction(F->getName());
```
Comment (jfix71) [inline lib/Graph/Graph.cpp]: nit: `auto *copyF`

File: lib/Graph/Graph.cpp [hunk 4]
```diff
@@ -4093,6 +4123,43 @@ static bool verifyNodeInput(const Node &N, size_t idx) {
...
+    currToNew[PH] = copyPH;
+  }
+  // Clone constants.
+  for (auto &C : getConstants()) {
+    auto *copyC = M->createConstant(C->getType(), C->getName(), C->getLayout());
+    copyC->assign(&C->getPayload());
+    currToNew[C] = copyC;
+  }
+  // Clone all functions.
+  for (auto &F : getFunctions()) {
```
Comment (jfix71) [inline lib/Graph/Graph.cpp]: nit: `auto *F` instead? Right now it's a reference to a pointer.

**Comment by opti-mix** [comment]
@jackm321 

> What is the expectation for modules that have been stripped? I think this will reallocate memory for the Constants but that might not be true. In any case, it would be good to document and test what the expected behavior is.

This is a great catch! I added a DCHECK for now to catch such cases where we have constants with unowned payloads and bail on it. These situations cannot be resolved by the cloner on its own without some input from the client.

File: tests/unittests/GraphTest.cpp [hunk 5]
```diff
@@ -674,6 +674,89 @@ TEST(Graph, functionCloneTest) {
...
+  // module was deleted.
+  EXPECT_TRUE(clonedM.verify());
+  // Compile and run the cloned model.
+  compileAndRun(clonedEE, clonedBindings, clonedM, "input", resultName);
+  // Store the result of running the cloned module.
+  clonedResult.assign(
+      clonedBindings.get(clonedBindings.getPlaceholderByName(resultName)));
+  // The results of execution should be exactly the same in both cases.
+  EXPECT_TRUE(originalResult.isEqual(clonedResult, 0));
+  clonedEE.setBackendName("Interpreter");
```
Comment (jfix71) [inline tests/unittests/GraphTest.cpp]: Why is this necessary?

File: lib/Graph/Graph.cpp [hunk 6]
```diff
@@ -4093,6 +4123,49 @@ static bool verifyNodeInput(const Node &N, size_t idx) {
...
+  for (auto &PH : getPlaceholders()) {
+    auto *copyPH = M->createPlaceholder(PH->getType(), PH->getName(),
+                                        PH->isTraining(), PH->getLayout());
+    currToNew[PH] = copyPH;
+  }
+  // Clone constants.
+  for (auto &C : getConstants()) {
+    // Cloner cannot decide on its own what to do with constants having unowned
+    // payloads. Some kind of policy/hook maybe needed in the future for
+    // decidiing what needs to be done in such cases.
```
Comment (jfix71) [inline lib/Graph/Graph.cpp]: typo: decidiing

File: tests/unittests/GraphTest.cpp (hunk 5 above)
Comment (opti-mix) [reply tests/unittests/GraphTest.cpp]: Oops. Good catch! This should not be there. It was from an experiment ;-) Removed!
